### ISSUE TITLE
Data preprocessing before training

### ISSUE THREAD

**Issue by jurajmuran321** [pjreddie/darknet#1947]
Is there any kind of data preprocessing before training? I did my own python implementation of Yolo in Keras. I tried to train it with data in darknet format. I resized all images to specified width and height, and used this way of annotation for every object:

class  center_x / im_width  center_y / im_height  width / im_width  height / im_height 

I managed to train it somehow, but it gives absolutely random predictions. The model and loss function are surely correct, so the problem might be with my data. However, I used the same data to succesfully train model with this C darknet implementation, and it did the prediction well. Can anyone help me? Thank you.

**Comment by mathemaphysics** [comment]
I'm in the same boat. In order to figure out how they normalize the images during preprocessing I needed to dig through their source code. Because I trained with my own data via darknet. And while `darknet detector test ...` works just fine, I can't get the output weights to produce good box results in OpenCV's DNN module using `readNetFromDarknet()` because I don't know their preprocessing scheme.

Eventually I stumbled on the fact that I had to use cv::IMREAD_UNCHANGED to load the image, get a blob from it by using scale factor of 0.00392 (or 1/255) while assuming the value of each pixel is CV_32F, i.e. 32-bit floating point. That scaling factor makes no sense at all to me in combination with CV_32F until I realized that CV_32F wasn't the RGB input size, but the blob output format, 32-bit float. I conclude that they're preprocessing to number between 0 and 1 and not telling us.
